### PR TITLE
openjdk17: init at 17.0.1+12

### PR DESCRIPTION
**Pull request by Uthar**
###### Motivation for this change

Package OpenJDK 17 LTS https://github.com/NixOS/nixpkgs/issues/137956

Notify maintainers: @edwtjo  @asbachb 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/openjdk/ignore-LegalNoticeFilePlugin.patch [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+-                // to contain the same content
+-                Optional<ResourcePoolEntry> ores =
+-                    entries.stream().filter(e -> e.linkedTarget() == null)
+-                           .findAny();
++            // if (errorIfNotSameContent) {
++            //     // all legal notices of the same file name are expected
++            //     // to contain the same content
++            //     Optional<ResourcePoolEntry> ores =
++            //         entries.stream().filter(e -> e.linkedTarget() == null)
++            //                .findAny();
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/openjdk/ignore-LegalNoticeFilePlugin.patch:18]:
```suggestion
```

makes the patch shorter and easier to understand

File: pkgs/development/compilers/openjdk/ignore-LegalNoticeFilePlugin.patch [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+-                if (ores.isPresent()) {
+-                    throw new PluginException(ores.get().path() + " " +
+-                        entry.path() + " contain different content");
+-                }
+-            }
++            //     if (ores.isPresent()) {
++            //         throw new PluginException(ores.get().path() + " " +
++            //             entry.path() + " contain different content");
++            //     }
++            // }
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/openjdk/ignore-LegalNoticeFilePlugin.patch:29]:
```suggestion
```

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -11722,27 +11727,28 @@ with pkgs;
...
 
-  jdk16 = openjdk16;
-  jdk16_headless = openjdk16_headless;
+  jdk17 = openjdk17;
+  jdk17_headless = openjdk17_headless;
 
   /* default JDK */
 
-  jdk = jdk16;
+  jdk-lts = jdk17;
```
Comment (xfix) [inline pkgs/top-level/all-packages.nix:11750]: I don't think introducing `jdk-lts` attribute is a good idea. If an application doesn't use internal APIs and doesn't parse its own bytecode then using `jdk` will work fine. If that's not the case then updating to newest JDK is going to break stuff. If an user wants an LTS, I think having to specify a version makes sense.

Comment (Uthar) [reply pkgs/top-level/all-packages.nix:11750]: I agree, the existing `jdk` alias itself seems good enough

**Comment by xfix** [comment]
Result of `nixpkgs-review pr 140257` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages marked as broken and skipped:</summary>
  <ul>
    <li>aldor</li>
    <li>foo-yc20</li>
    <li>junit</li>
    <li>junixsocket</li>
    <li>mindustry</li>
    <li>mindustry-server</li>
    <li>mindustry-wayland</li>
    <li>octave-jit</li>
    <li>octavePackages.fem-fenics</li>
    <li>octavePackages.vibes</li>
    <li>octavePackages.vrml</li>
    <li>openfire</li>
    <li>purePackages.faust</li>
    <li>python38Packages.skein</li>
    <li>python39Packages.skein</li>
  </ul>
</details>
<details>
  <summary>352 packages failed to build:</summary>
  <ul>
    <li>DisnixWebService</li>
    <li>adobe_flex_sdk (apache-flex-sdk)</li>
    <li>agi</li>
    <li>alchemy</li>
    <li>alda</li>
    <li>amidst</li>
    <li>ammonite (ammonite-repl)</li>
    <li>ammonite_2_12</li>
    <li>android-backup-extractor</li>
    <li>apache-directory-studio</li>
    <li>apache-jena</li>
    <li>fuseki (apache-jena-fuseki ,joseki)</li>
    <li>apksigcopier</li>
    <li>apksigner</li>
    <li>apktool</li>
    <li>archi</li>
    <li>arduino</li>
    <li>arduino-core (arduino_core)</li>
    <li>asciidoc-full-with-plugins</li>
    <li>asciidoctorj</li>
    <li>aspectj</li>
    <li>astrolabe-generator</li>
    <li>atlassian-cli</li>
    <li>avro-tools</li>
    <li>aws_mturk_clt</li>
    <li>awstats</li>
    <li>axis2</li>
    <li>basex</li>
    <li>bfg-repo-cleaner</li>
    <li>bftools</li>
    <li>bitwig-studio1</li>
    <li>bitwig-studio2</li>
    <li>bloop</li>
    <li>bonnmotion</li>
    <li>boot</li>
    <li>briss</li>
    <li>ccemux</li>
    <li>cfr</li>
    <li>checkstyle</li>
    <li>clooj</li>
    <li>closurecompiler</li>
    <li>confluent-platform</li>
    <li>coursier</li>
    <li>crowdin-cli</li>
    <li>cryptomator</li>
    <li>davmail</li>
    <li>dbqn</li>
    <li>dcs</li>
    <li>dex2jar</li>
    <li>diffoscope</li>
    <li>ditaa</li>
    <li>domination</li>
    <li>ec2_api_tools</li>
    <li>eddy</li>
    <li>emem</li>
    <li>emscripten</li>
    <li>emscriptenPackages.json_c</li>
    <li>emscriptenPackages.libxml2</li>
    <li>emscriptenPackages.xmlmirror</li>
    <li>emscriptenPackages.zlib</li>
    <li>epubcheck</li>
    <li>ergo</li>
    <li>exhibitor</li>
    <li>fabric-installer</li>
    <li>faust</li>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jack</li>
    <li>faust2jackrust</li>
    <li>faust2jaqt</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>faustPhysicalModeling</li>
    <li>faustStk</li>
    <li>faustlive</li>
    <li>fdroidserver</li>
    <li>flexibee</li>
    <li>flink</li>
    <li>flyway</li>
    <li>frostwire</li>
    <li>frostwire-bin</li>
    <li>gama</li>
    <li>ganttproject-bin</li>
    <li>geogebra</li>
    <li>geoipjava</li>
    <li>gephi</li>
    <li>global-platform-pro</li>
    <li>gogui</li>
    <li>google-java-format</li>
    <li>gpsprune</li>
    <li>groove</li>
    <li>groovy</li>
    <li>h2</li>
    <li>haxePackages.hxjava</li>
    <li>hdfview</li>
    <li>hsqldb</li>
    <li>html5validator</li>
    <li>hydraAntLogger</li>
    <li>imagej</li>
    <li>ino</li>
    <li>ipscan</li>
    <li>jadx</li>
    <li>jameica</li>
    <li>java-language-server</li>
    <li>jbake</li>
    <li>jbang</li>
    <li>jd-gui</li>
    <li>jdiskreport</li>
    <li>jdk</li>
    <li>jdk17_headless</li>
    <li>jedit</li>
    <li>jenkins</li>
    <li>jflex</li>
    <li>jibri</li>
    <li>jicofo</li>
    <li>jing (trang)</li>
    <li>jitsi-videobridge</li>
    <li>jmeter</li>
    <li>jmusicbot</li>
    <li>jnetmap</li>
    <li>josm</li>
    <li>jre_minimal</li>
    <li>jruby</li>
    <li>jsvc</li>
    <li>jugglinglab</li>
    <li>jython</li>
    <li>kapitonov-plugins-pack</li>
    <li>key</li>
    <li>keycloak</li>
    <li>kotlin</li>
    <li>ktlint</li>
    <li>languagetool</li>
    <li>leiningen</li>
    <li>leo3-bin</li>
    <li>libreoffice (libreoffice-still)</li>
    <li>libreoffice-fresh</li>
    <li>libreoffice-fresh-unwrapped</li>
    <li>libreoffice-qt</li>
    <li>libreoffice-still-unwrapped</li>
    <li>librsb</li>
    <li>liquibase</li>
    <li>logisim</li>
    <li>logisim-evolution</li>
    <li>lombok</li>
    <li>lumo</li>
    <li>ma1sd</li>
    <li>macse</li>
    <li>magnetophonDSP.CharacterCompressor</li>
    <li>magnetophonDSP.CompBus</li>
    <li>magnetophonDSP.ConstantDetuneChorus</li>
    <li>magnetophonDSP.LazyLimiter</li>
    <li>magnetophonDSP.MBdistortion</li>
    <li>magnetophonDSP.RhythmDelay</li>
    <li>magnetophonDSP.VoiceOfFaust</li>
    <li>magnetophonDSP.faustCompressors</li>
    <li>magnetophonDSP.pluginUtils</li>
    <li>magnetophonDSP.shelfMultiBand</li>
    <li>makemkv</li>
    <li>mars-mips</li>
    <li>martyr</li>
    <li>marvin</li>
    <li>maven</li>
    <li>metals</li>
    <li>micronaut</li>
    <li>minc_widgets</li>
    <li>minecraft</li>
    <li>minecraft-server</li>
    <li>mkgmap</li>
    <li>mkgmap-splitter</li>
    <li>mooSpace</li>
    <li>msgviewer</li>
    <li>multimc</li>
    <li>munin</li>
    <li>nailgun</li>
    <li>netlogo</li>
    <li>octave</li>
    <li>octaveFull</li>
    <li>octavePackages.arduino</li>
    <li>octavePackages.audio</li>
    <li>octavePackages.bim</li>
    <li>octavePackages.bsltl</li>
    <li>octavePackages.cgi</li>
    <li>octavePackages.communications</li>
    <li>octavePackages.control</li>
    <li>octavePackages.data-smoothing</li>
    <li>octavePackages.database</li>
    <li>octavePackages.dataframe</li>
    <li>octavePackages.dicom</li>
    <li>octavePackages.divand</li>
    <li>octavePackages.doctest</li>
    <li>octavePackages.econometrics</li>
    <li>octavePackages.financial</li>
    <li>octavePackages.fits</li>
    <li>octavePackages.fpl</li>
    <li>octavePackages.fuzzy-logic-toolkit</li>
    <li>octavePackages.ga</li>
    <li>octavePackages.general</li>
    <li>octavePackages.generate_html</li>
    <li>octavePackages.geometry</li>
    <li>octavePackages.gsl</li>
    <li>octavePackages.image</li>
    <li>octavePackages.image-acquisition</li>
    <li>octavePackages.instrument-control</li>
    <li>octavePackages.interval</li>
    <li>octavePackages.io</li>
    <li>octavePackages.level-set</li>
    <li>octavePackages.linear-algebra</li>
    <li>octavePackages.lssa</li>
    <li>octavePackages.ltfat</li>
    <li>octavePackages.mapping</li>
    <li>octavePackages.matgeom</li>
    <li>octavePackages.miscellaneous</li>
    <li>octavePackages.msh</li>
    <li>octavePackages.mvn</li>
    <li>octavePackages.nan</li>
    <li>octavePackages.ncarray</li>
    <li>octavePackages.netcdf</li>
    <li>octavePackages.nurbs</li>
    <li>octavePackages.ocl</li>
    <li>octavePackages.octclip</li>
    <li>octavePackages.octproj</li>
    <li>octavePackages.optics</li>
    <li>octavePackages.optim</li>
    <li>octavePackages.optiminterp</li>
    <li>octavePackages.parallel</li>
    <li>octavePackages.quaternion</li>
    <li>octavePackages.queueing</li>
    <li>octavePackages.signal</li>
    <li>octavePackages.sockets</li>
    <li>octavePackages.sparsersb</li>
    <li>octavePackages.splines</li>
    <li>octavePackages.statistics</li>
    <li>octavePackages.stk</li>
    <li>octavePackages.strings</li>
    <li>octavePackages.struct</li>
    <li>octavePackages.symbolic</li>
    <li>octavePackages.tisean</li>
    <li>octavePackages.tsa</li>
    <li>octavePackages.video</li>
    <li>octavePackages.windows</li>
    <li>octavePackages.zeromq</li>
    <li>omegat</li>
    <li>opa</li>
    <li>open-ecard</li>
    <li>openapi-generator-cli</li>
    <li>openapi-generator-cli-unstable</li>
    <li>openems</li>
    <li>opengrok</li>
    <li>openjump</li>
    <li>openlp</li>
    <li>openlpFull</li>
    <li>optifine</li>
    <li>papermc</li>
    <li>perl532Packages.NetZooKeeper</li>
    <li>perl534Packages.NetZooKeeper</li>
    <li>picard-tools</li>
    <li>picolisp</li>
    <li>plantuml</li>
    <li>plantuml-server</li>
    <li>plm</li>
    <li>polylith</li>
    <li>polymake</li>
    <li>portmod</li>
    <li>prometheus-jmx-httpserver</li>
    <li>python38Packages.databricks-connect</li>
    <li>python38Packages.hydra</li>
    <li>python38Packages.omegaconf</li>
    <li>python38Packages.python-csxcad</li>
    <li>python38Packages.python-openems</li>
    <li>python38Packages.sphinxcontrib-plantuml (python38Packages.sphinxcontrib_plantuml)</li>
    <li>python39Packages.databricks-connect</li>
    <li>python39Packages.hydra</li>
    <li>python39Packages.omegaconf</li>
    <li>python39Packages.python-csxcad</li>
    <li>python39Packages.python-openems</li>
    <li>python39Packages.sphinxcontrib-plantuml (python39Packages.sphinxcontrib_plantuml)</li>
    <li>quantomatic</li>
    <li>rascal</li>
    <li>riemann</li>
    <li>runelite</li>
    <li>saxon</li>
    <li>saxon-he</li>
    <li>saxonb</li>
    <li>saxonb_9_1</li>
    <li>sbt</li>
    <li>sbt-extras</li>
    <li>sbt-with-scala-native</li>
    <li>scalafmt</li>
    <li>scenic-view</li>
    <li>schemaspy</li>
    <li>selendroid</li>
    <li>selenium-server-standalone</li>
    <li>seyren</li>
    <li>shattered-pixel-dungeon</li>
    <li>signald</li>
    <li>sleuthkit</li>
    <li>smc</li>
    <li>snapdragon-profiler</li>
    <li>snpeff</li>
    <li>solr (solr_8)</li>
    <li>sonar-scanner-cli</li>
    <li>sourcetrail</li>
    <li>spring</li>
    <li>spring-boot (spring-boot-cli)</li>
    <li>springLobby</li>
    <li>squirrel-sql</li>
    <li>sshoogr</li>
    <li>storm</li>
    <li>subgit</li>
    <li>subsonic</li>
    <li>sumo</li>
    <li>swagger-codegen</li>
    <li>swagger-codegen3</li>
    <li>swingsane</li>
    <li>swt</li>
    <li>tabula</li>
    <li>tabula-java</li>
    <li>tambura</li>
    <li>teensyduino</li>
    <li>tessera</li>
    <li>timedoctor</li>
    <li>tomcat-native</li>
    <li>tomcat_connectors</li>
    <li>tr-patcher</li>
    <li>turbovnc</li>
    <li>tvbrowser-bin</li>
    <li>umlet</li>
    <li>unciv</li>
    <li>unoconv</li>
    <li>varscan</li>
    <li>vassal</li>
    <li>vimPlugins.lens-vim</li>
    <li>vimPlugins.vim-grammarous</li>
    <li>visualvm</li>
    <li>vscode-extensions.redhat.java</li>
    <li>vue</li>
    <li>vuze</li>
    <li>weka</li>
    <li>welkin</li>
    <li>workcraft</li>
    <li>xmind</li>
    <li>xmloscopy</li>
    <li>yed</li>
    <li>yuicompressor</li>
    <li>zap</li>
    <li>zgrviewer</li>
    <li>zipkin</li>
    <li>zkfuse</li>
    <li>zookeeper_mt</li>
    <li>zxing</li>
  </ul>
</details>

**Comment by Uthar** [comment]
> Result of `nixpkgs-review pr 140257` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 
> 15 packages marked as broken and skipped:
> 352 packages failed to build:

Patch went bad, pushed a fix

**Comment by erasmas** [comment]
Looking forward for this package to get merged. 🤞🏼

**Review by VergeDX — COMMENTED** [review]
LGTM

**Comment by pmeiyu** [comment]
@Uthar After that, you can do a force push to your own branch.

By the way, I like your work.  Thank you for this PR.

**Comment by VergeDX** [comment]
Always remember do `git commit --amend` : )
GitHub will still track your history : )

**Comment by asbachb** [comment]
Just FYI there's a new release of java 17.0.1: https://github.com/openjdk/jdk17u/tags

**Comment by pmeiyu** [comment]
> Just FYI there's a new release of java 17.0.1: https://github.com/openjdk/jdk17u/tags

Yes. We should use the `jdk-17.0.1-ga` tag.  `jdk-17+35` seems too old.

**Comment by Uthar** [comment]
Bumped it to the latest tag

**Comment by gvolpe** [comment]
Exciting work, can't wait for this to be merged! I'm already using it in a private repo and seems to work fine, thanks a lot for putting up the effort @Uthar 🤩

**Review by gvolpe — APPROVED** [review]

**Comment by Uthar** [comment]
Thanks everyone for help with this package :)

**Pull request merged**